procedures.sql

```diff
@@ -0,0 +1,1136 @@
+GO
+-- EXEC SetInternshipExamResult 17, 23, 60
+
+
+
+CREATE PROCEDURE LogMeetingAttendance
+    @student_id INT,
+    @meeting_id INT
+AS
+BEGIN
+    INSERT INTO StudentMeetingAttendance (student_id, meeting_id)
+    VALUES (@student_id, @meeting_id);
+END;
+GO
+-- EXEC LogMeetingAttendance 17, 23
+
+
+
+CREATE PROCEDURE AddParameter
+    @name NVARCHAR(50),
+    @value NVARCHAR(50),
+    @date DATE = NULL
+AS
+BEGIN
+    IF @date IS NULL
+        SET @date = CURRENT_TIMESTAMP;
+
+    INSERT INTO Parameter (name, value, date)
+    VALUES (@name, @value, @date);
+END;
+GO
+-- EXEC AddParameter 'TEST PARAMETER', 2, '2019-01-01';
+-- EXEC AddParameter 'TEST PARAMETER', 'value';
+
+
+
+CREATE PROCEDURE CreateBasket
+@student_id INT
+AS
+BEGIN
+    INSERT INTO Basket (student_id, state, create_date)
+    VALUES (@student_id, 'open', CURRENT_TIMESTAMP);
+END;
+GO
+-- EXEC CreateBasket 17
+
+CREATE PROCEDURE AddItemToBasket
+    @student_id INT,
+    @course_id INT = NULL,
+    @meeting_id INT = NULL,
+    @studies_id INT = NULL,
+    @webinar_id INT = NULL
+AS
+BEGIN
+    DECLARE @basket_id INT = dbo.getStudentsBasket(@student_id);
+
+    IF @basket_id IS NULL
+        BEGIN
+            EXEC CreateBasket @student_id;
+            SET @basket_id = dbo.getStudentsBasket(@student_id);
+        END
+
+    IF @course_id IS NOT NULL
+        BEGIN
+            INSERT INTO BasketItem (basket_id, course_id)
+            VALUES (@basket_id, @course_id);
+        END
+
+    IF @meeting_id IS NOT NULL
+        BEGIN
+            INSERT INTO BasketItem (basket_id, meeting_id)
+            VALUES (@basket_id, @meeting_id);
+        END
+
+    IF @studies_id IS NOT NULL
+        BEGIN
+            INSERT INTO BasketItem (basket_id, studies_id)
+            VALUES (@basket_id, @studies_id);
+        END
+
+    IF @webinar_id IS NOT NULL
+        BEGIN
+            INSERT INTO BasketItem (basket_id, webinar_id)
+            VALUES (@basket_id, @webinar_id);
+        END
+END;
+GO
+-- EXEC AddItemToBasket 17, @course_id = 3
+
+CREATE PROCEDURE RemoveItemFromBasket
+    @basket_id INT,
+    @course_id INT = NULL,
+    @meeting_id INT = NULL,
+    @studies_id INT = NULL,
+    @webinar_id INT = NULL
+AS
+BEGIN
+    IF @course_id IS NOT NULL
+        BEGIN
+            DELETE FROM BasketItem
+            WHERE basket_id = @basket_id AND course_id = @course_id;
+        END
+
+    IF @meeting_id IS NOT NULL
+        BEGIN
+            DELETE FROM BasketItem
+            WHERE basket_id = @basket_id AND meeting_id = @meeting_id;
+        END
+
+    IF @studies_id IS NOT NULL
+        BEGIN
+            DELETE FROM BasketItem
+            WHERE basket_id = @basket_id AND studies_id = @studies_id;
+        END
+
+    IF @webinar_id IS NOT NULL
+        BEGIN
+            DELETE FROM BasketItem
+            WHERE basket_id = @basket_id AND webinar_id = @webinar_id;
+        END
+END;
+GO
+-- EXEC RemoveItemFromBasket 35, @course_id = 3
+
+CREATE PROCEDURE FailBasketPayment
```

Comment:
`register_failure`... Do I need to explain again?

Reply:
That would be kind of you

Reply:
OK. You don't fail the basket payment, but register such a fact.